server/asset/btc/rpcclient.go

```diff
@@ -130,8 +164,195 @@ func (rc *RPCClient) GetRawTransactionVerbose(txHash *chainhash.Hash) (*btcjson.
+}
+
+// MedianFeeRate returns the median rate from the specified block.
+func (rc *RPCClient) medianFeeRate() (uint64, error) {
+	blockHash, err := rc.GetBestBlockHash()
+	if err != nil {
+		return 0, err
+	}
+
+	res := struct {
+		FeeRatePercentiles []uint64 `json:"feerate_percentiles"` // 10th, 25th, 50th, 75th, and 90th percentiles
+		TxCount            int      `json:"txs"`
+	}{}
+
+	categories := []string{"feerate_percentiles", "txs"}
+
+	// We need at least a few transactions, but there's nothing stopping a miner
+	// from publishing empty blocks, regardless of the current mempool state,
+	// and we would want to reduce the impact of a particularly choosy node as
+	// well, I think. So we'll check > 100 transaction in up to 10 blocks,
+	// taking a weighted average of the fees. Consider two cases.
+	//
+	// 1) When the first block has > 100 transactions, it probably indicates the
+	// the miner is operating as expected and the blockchain is busy. In this
+	// case, the most recent block is the best estimate, since fees from older
+	// blocks could become quickly outdated on a busy chain.
+	//
+	// 2) If the first block has fewer transactions, it may be safe to say that
+	// either a) the miner is not stuffing blocks as expected or b) the
+	// blockchain does not have 100 txs/block worth of traffic. Because we have
+	// no historical view of mempool, it's impossible to say which one it is,
+	// though. In this case, taking a weighted average over a few recent blocks
+	// would provide a better estimate.
+
+	var blocksChecked, txCount int
+	var weight uint64
+	for txCount < 101 {
+		if blocksChecked >= rc.maxFeeBlocks {
+			return 0, ErrNoCompetition
+		}
+
+		if err := rc.call(methodGetBlockStats, anylist{blockHash.String(), categories}, &res); err != nil {
+			return 0, err
+		}
+		if len(res.FeeRatePercentiles) != 5 {
+			return 0, fmt.Errorf("unexpected feerate_percentiles response. %d entries", len(res.FeeRatePercentiles))
+		}
+
+		feeRate := res.FeeRatePercentiles[2]
+
+		weight += uint64(res.TxCount) * feeRate
+
+		txCount += res.TxCount
+		blocksChecked++
+
+		if txCount >= 101 {
+			break
+		}
+
+		// Not enough transactions to count yet.
+		verboseBlock, err := rc.GetBlockVerbose(blockHash)
+		if err != nil {
+			return 0, err
+		}
+
+		blockHash, err = chainhash.NewHashFromStr(verboseBlock.PreviousHash)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	// rounded average
+	return uint64(math.Round(float64(weight) / float64(txCount))), nil
+}
+
+// medianFeesTheHardWay calculates the median fees from the previous block(s).
+// medianFeesTheHardWay is used for assets that don't have a getblockstats RPC,
+// and is only useful for non-segwit assets.
+func (rc *RPCClient) medianFeesTheHardWay() (uint64, error) {
+	const numTxs = 101
+
+	iHash, err := rc.GetBestBlockHash()
+	if err != nil {
+		return 0, err
+	}
+
+	txs := make([]*wire.MsgTx, 0, numTxs)
+	prevOuts := make(map[chainhash.Hash]map[int]int64, numTxs)
```

Comment:
pls doc the keys and values: first map key is prevout hash (funding tx), the second map is prev out idx => value